File: win/Sources/ScreenOverlay.cs
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;

namespace GlassFace;

/// <summary>
/// One display's overlay: the transparent window plus the mirrored camera image drawn into
/// it. The image is shown at reduced opacity (the "glass" effect) and horizontally flipped
/// for a natural selfie view, matching the macOS preview layer's mirroring.
/// </summary>
public sealed class ScreenOverlay
{
    private readonly OverlayWindow _window;
    private readonly Image _image;

    public ScreenOverlay(NativeMethods.RECT bounds, float opacity)
    {
        _window = new OverlayWindow(bounds);

        _image = new Image
        {
            Stretch = Stretch.UniformToFill,            // ≈ AVLayerVideoGravityResizeAspectFill
            Opacity = opacity,
            RenderTransformOrigin = new Point(0.5, 0.5),
            RenderTransform = new ScaleTransform(-1, 1)  // mirror (selfie view)
        };
        RenderOptions.SetBitmapScalingMode(_image, BitmapScalingMode.LowQuality); // fast preview scaling

        var root = new Grid { Background = Brushes.Transparent };
        root.Children.Add(_image);
        _window.Content = root;
    }

    public void Present() => _window.Show();

    public void SetOpacity(float value) => _image.Opacity = value;

    public void SetSource(ImageSource source) => _image.Source = source;
}

File: win/Sources/ScreenOverlay.cpp
#include "pch.h"
#include "ScreenOverlay.h"
#include "OverlayWindow.h"

using namespace winrt;
using namespace winrt::Microsoft::UI::Xaml;
using namespace winrt::Microsoft::UI::Xaml::Controls;
using namespace winrt::Microsoft::UI::Xaml::Media;
using namespace winrt::Microsoft::UI::Xaml::Media::Imaging;

namespace glassface
{
    ScreenOverlay::ScreenOverlay(RECT const& bounds, float opacity, SoftwareBitmapSource const& source)
    {
        m_window = Window();

        Grid root;
        root.Background(SolidColorBrush(winrt::Microsoft::UI::Colors::Transparent()));

        m_image = Image();
        m_image.Stretch(Stretch::UniformToFill);          // ≈ AVLayerVideoGravityResizeAspectFill
        m_image.Opacity(opacity);
        m_image.Source(source);
        m_image.RenderTransformOrigin(winrt::Windows::Foundation::Point{ 0.5f, 0.5f });

        ScaleTransform flip;                              // mirror (selfie view)
        flip.ScaleX(-1.0);
        flip.ScaleY(1.0);
        m_image.RenderTransform(flip);

        root.Children().Append(m_image);
        m_window.Content(root);

        ConfigureOverlayWindow(m_window, bounds);
    }

    void ScreenOverlay::Present()
    {
        m_window.Activate();
    }

    void ScreenOverlay::SetOpacity(float value)
    {
        m_image.Opacity(value);
    }
}
